src/ledger/insert_shred.zig

```diff
@@ -810,6 +873,12 @@ pub const ShredInserter = struct {
         data_index: *meta.ShredIndex,
     ) bool {
         const shred_index: u64 = @intCast(shred.fields.common.index);
+        // QUESTION: is the "shred_index < slot_meta.consecutive_received_from_0" check needed?
+        // Ie would there ever be a case where data_index.contains(shred_index) is true but
+        // shred_index < slot_meta.consecutive_received_from_0 is false? If not then is data_index.contains(shred_index)
+        // not enough check?
+        // Also if shred_index > slot_meta.consecutive_received_from_0 is true and also data_index.contains(shred_index) is true
+        // is it a valid case case for data_index.contains(shred_index) to be true but shred_index > slot_meta.consecutive_received_from_0.
         return shred_index < slot_meta.consecutive_received_from_0 or
             data_index.contains(shred_index);
```

Comment:
You made a great observation here. But there is actually a specific performance reason for this redundancy.

The idea is that `<` is faster than `.contains()`. We run the fast check first. If true, that's a definitive answer. But if false, we can't know for sure. This comparison could have false negatives. So in that case, we run the slower check to get a definitive answer.